client/cli/src/config.rs

```diff
@@ -255,17 +255,31 @@ pub trait CliConfiguration: Sized {
 	/// Get the RPC HTTP address (`None` if disabled).
 	///
 	/// By default this is `None`.
-	fn rpc_http(&self) -> Result<Option<SocketAddr>> {
+	fn rpc_http(&self, _default_port: u16) -> Result<Option<SocketAddr>> {
 		Ok(Default::default())
 	}
 
+	/// Get the default RPC HTTP port.
+	///
+	/// By default this is 9933.
+	fn default_rpc_http_port(&self) -> Result<u16> {
+		Ok(9933)
+	}
```

Comment:
Does this need to return a `Result`?

Reply:
:thinking: the idea with all the results is that it allows the user to put a behavior that can fail.

But for the `_params` method we don't use Result because it makes no sense.

Here it's the first time we use methods to get the default values (the other default values are in `create_configuration` directly).

I don't see why someone would want their default value to fail. But then, is `&self` a useful parameter for a method that returns a default value?

I think it would be best to remove the Result and the `&self` altogether. Let me know if that sounds good/logical to you.

Reply:
I think that removing the `Result` and `&self` is best for now. If a team building on substrate presents a good reason to allow returning an error in this function then we can always allow that later.

Reply:
Turns out I can't make the variants for the subcommands anymore: https://github.com/paritytech/substrate/pull/6250/commits/c7f985129b956c29ccb1a7bbd6f4e81195f5ef13

I guess that makes sense... :thinking: it wouldn't make much sense to change the default values per subcommand.

I'm starting to wonder if it wouldn't be best to have this kind of default values on the SubstrateCli trait instead of ConfigurationCli.

Reply:
Oh, well maybe taking `&self` makes sense then.

Reply:
That kind of default should be valid for a whole program and not customized somehow per subcommand/runcmd. So I think it is better without the `&self` but right now it is badly implemented because it can vary per subcommand. So I'm thinking about moving it to SubstrateCli along the other statics that define what the node is.

The alternative would be to keep the `&self` but not the `Result` but then it means it is still possible to have a different default value per subcommand and also the lack of Result makes it inconsistent with the other methods (maybe rightfully so).

I will move it to SubstrateCli if you don't have any objection. What we can do with them will be more limited but I think it will be less surprising and more consistent overall.